fix(explore): Chart save modal displays error instead of failing silently

kgabryje opened the pull request:
### SUMMARY
When user tried to save a chart and a request executed by SaveModal chart failed, the modal would just close and there would be no indication that an error occurred. That might lead the user to think that their chart was successfully saved even though it wasn't.
This PR fixes that behaviour by not closing the modal and displaying an alert with error message within it. 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before:

https://user-images.githubusercontent.com/15073128/197514948-a996021b-bd5d-4939-aee5-c9f23c34b2d6.mov

After:

https://user-images.githubusercontent.com/15073128/197514625-2667b4fe-519b-4313-be65-58e00a31c66f.mov

### TESTING INSTRUCTIONS
1. Open the chart save modal
2. Turn off the Internet (to trigger an error)
3. Click save button 
4. See that the error is displayed and the modal doesn't close

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

CC @codyml

kgabryje committed:
fix(explore): Chart save modal displays error instead of failing silently

kgabryje commented:
/testenv up

codyml commented on superset-frontend/src/explore/components/SaveModal.tsx:
```diff
@@ -150,9 +151,6 @@ class SaveModal extends React.Component<SaveModalProps, SaveModalState> {
     this.props.actions.removeSaveModalAlert();
-    // eslint-disable-next-line @typescript-eslint/no-unused-vars
-
-    let promise = Promise.resolve();
```
Changing promises to `await`s looks great, thanks!  Two related notes:
1. Since all the `catch` blocks are the same, should we just wrap the entire function body in one `try` block?
2. Previously, since these promises had no `catch` block, errors would go unhandled and therefore show up in the browser console.  Should we preserve that behavior by re-throwing or not catching errors?

If you think yes for both of those, we could just do something like:
```javascript
async saveOrOverwrite(goodish:boolean) {
  try {
    // If anything throws here, modal will stay open, finally block will execute,
    // and error will remain unhandled because there's no catch block
    this.props.onHide();
  } finally {
    this.setState({ isLoading: false });
  }
}

kgabryje replied:
Great idea! Done

codyml reviewed (commented):
Thanks!  I left one non-blocking comment about try/catch.  The loading indicator also looks a little strange to me left-justified.  Would it be possible to make it centered?

geido commented:
Hey @kgabryje I think I am noticing a weird behaviour when saving the chart as a new one. After saving the chart shows the empty state.

https://user-images.githubusercontent.com/60598000/197768634-56288be1-d1d8-40c4-80c7-6eb7d5f57c27.mp4

kgabryje committed:
Move reading saveAction from url to redux

kgabryje committed:
small lint changes

kgabryje commented:
@geido @codyml Thank you for your feedback and testing! I fixed the issue with chart disappearing after saving.
The problem was a race condition between re-hydrating explore after saving and triggering query. Because of some strange flows in the code, delaying closing the save modal was causing the query to be triggered before Explore was rehydrated. The result was that the query was updating the previous chart (previous slice_id), instead of the freshly created one.
I fixed this issue by moving the logic responsible for detecting saving from URL (we were reading `saveaction` search parameter in the URL to decide if query needs to be triggered) to Redux (which is now being set by `hydrateExplore`, so now the correct order of actions is ensured).


https://user-images.githubusercontent.com/15073128/197980861-9190ccc0-9dfe-4a71-9778-ba481d75fc9c.mov

kgabryje committed:
Wrap the whole saveOrOverwrite in try/finally

kgabryje committed:
Fix ut

kgabryje committed:
Fix ut

codyml reviewed (commented):
Amazing job fixing the chart refresh error!!  Code looks great; two things:
- I think the loading indicator is still left-justified
- I noticed that the overwrite/saveas toggle is now sticky on the same chart, so if you open a chart that you don't have permission to overwrite and save it as a new chart, the next time you go to save it'll still be on Save As, whereas previously I think it would always default to Overwrite if you had permissions.  Do we want to maintain that previous behavior?

kgabryje committed:
Fix loader position

kgabryje committed:
Fix resetting radio button state after closing the modal

kgabryje commented:
> * I think the loading indicator is still left-justified

Oops, I didn't notice your first comment. Fixed 🙂 

> * I noticed that the overwrite/saveas toggle is now sticky on the same chart, so if you open a chart that you don't have permission to overwrite and save it as a new chart, the next time you go to save it'll still be on Save As, whereas previously I think it would always default to Overwrite if you had permissions.  Do we want to maintain that previous behavior?

Well spotted! 🙏 Fixed

codyml reviewed (approved):
Tested locally, LGTM!

rusackas reviewed (approved):
LGTM!

kgabryje commented:
/testenv up
